### PR TITLE
[SPARK-50286][SQL] Correctly propagate SQL options to WriteBuilder

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

SPARK-49098 introduced a SQL syntax to allow users to set table options on DSv2 write cases, but unfortunately, the options set by SQL are not propagated correctly to the underlying DSv2 `WriteBuilder`

```
INSERT INTO $t1 WITH (`write.split-size` = 10) SELECT ...
```

```
df.writeTo(t1).option("write.split-size", "10").append()
```

From the user's perspective, the above two are equivalent, but internal implementations differ slightly. Both of them are going to construct an

```
AppendData(r: DataSourceV2Relation, ..., writeOptions, ...)
```

but the SQL `options` are carried by `r.options`, and the `DataFrame` API `options` are carried by `writeOptions`. Currently, only the latter is propagated to the `WriteBuilder`, and the former is silently dropped. This PR fixes the above issue by merging those two `options`.

Currently, the `options` propagation is inconsistent in `DataFrame`, `DataFrameV2`, and SQL:
- DataFrame API, the same `options` are carried by both `writeOptions` and `DataSourceV2Relation`
- DataFrameV2 API cases, options are only carried by `write options`
- SQL, `options` are only carried by `DataSourceV2Relation`

BTW, `SessionConfigSupport` only takes effect on `DataFrame` and `DataFrameV2` API, it is not considered in the `SQL` read/write path entirely in the current codebase.

### Why are the changes needed?

Correctly propagate SQL options to `WriteBuilder`, to complete the feature added in SPARK-49098, so that DSv2 implementations like Iceberg can benefit.

### Does this PR introduce _any_ user-facing change?

No, it's an unreleased feature.

### How was this patch tested?

UTs added by SPARK-36680 and SPARK-49098 are updated also to check SQL `options` are correctly propagated to the physical plan

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by pan3793** [comment]
cc @szehon-ho @cloud-fan @dongjoon-hyun

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala [hunk 1]
```diff
@@ -44,7 +46,7 @@ object V2Writes extends Rule[LogicalPlan] with PredicateHelper {
 
   override def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
     case a @ AppendData(r: DataSourceV2Relation, query, options, _, None, _) =>
-      val writeBuilder = newWriteBuilder(r.table, options, query.schema)
+      val writeBuilder = newWriteBuilder(r.table, r.options.asScala.toMap ++ options, query.schema)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:49]: can we add an assert that only one of them can be non empty?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+package org.apache.spark.sql.connector
+
+import org.apache.spark.sql.{AnalysisException, Row}
+import org.apache.spark.sql.catalyst.plans.logical.{AppendData, CommandResult, OverwriteByExpression}
+import org.apache.spark.sql.connector.catalog.InMemoryBaseTable
+import org.apache.spark.sql.execution.CommandResultExec
+import org.apache.spark.sql.execution.datasources.v2._
+
+class DataSourceV2OptionSuite extends DatasourceV2SQLBase {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala:27]:
```suggestion
class DataSourceV2OptionSQLSuite extends DatasourceV2SQLBase {
```

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala:27]: since this is testing SQL API only.

**Review by cloud-fan — APPROVED** [review]
This is a good catch!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for the fix, @pan3793 .

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+          assert(append.info.options.get("write.split-size") === "10")
+      }
+      assert (collected.size == 1)
+
+      val insertResult = sql(s"SELECT * FROM $t1")
+      checkAnswer(insertResult, Seq(Row(1, "a"), Row(2, "b")))
+    }
+  }
+
+  test("SPARK-36680, SPARK-50286: Supports Dynamic Table Options for SQL Insert Overwrite") {
```
Comment (szehon-ho) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala:92]: This is my fault, but we can optionally change the first JIRA's in these tests to SPARK-49098 as its the one that added the support to the inserts?

**Review by szehon-ho — COMMENTED** [review]
Thanks, did not realize this.  Looks from @cloud-fan comment that only one can be set?

**Comment by pan3793** [comment]
> ... that only one can be set

@szehon-ho yes, as mentioned in the description, DataFrame's API `options` go `writeOptions` while SQL `options` go `r.options`, I think they won't be set together in normal cases, but it would be great if someone could double check that.

**Comment by pan3793** [comment]
Wait, I forget the [`SessionConfigSupport`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SessionConfigSupport.java)

In fact, I submitted a [PR](https://github.com/apache/iceberg/pull/7732) to Iceberg a patch to support this feature, but unfortunately, this patch doesn't seem to be getting attention, @szehon-ho do you think we can re-open this PR and get it in? If so, the assumption would not hold.

> ... that only one can be set

and we should define the priority, I think it should be

1. `options` from SQL `options`
2. `options` from DataFrame API
3. `options` from session configuration

Currently, if there are duplicated options, 2 overrides 3, see https://github.com/apache/spark/blob/c1968a1bd85fa6f1a274849427105fa5327ca7e0/sql/core/src/main/scala/org/apache/spark/sql/internal/DataFrameWriterImpl.scala#L141-L142

@cloud-fan, do you think the proposed priority makes sense? or any new ideas?

**Comment by cloud-fan** [comment]
yea 3 should have lower priority.

**Comment by szehon-ho** [comment]
I think assert as @cloud-fan suggest makes sense if no known usecase of 1+2 together?  (can see if anything fails).  

Yea sure I re-opened the https://github.com/apache/iceberg/pull/7732 and will take a look.  Yea makes sense for these confs to have lower priority vs the SQL/DF ones, as there's one per entire source.

**Comment by pan3793** [comment]
I spent a few hours to modify the test cases but they do not work as expected, it seems `SessionConfigSupport` was not considered in the SQL read/write path entirely in the current codebase, though it should be, according to `SessionConfigSupport`'s docs

> Data sources can implement this interface to propagate session configs with the specified key-prefix to all data source operations in this session.

Making `SessionConfigSupport` work on the SQL read/write path may be out of the scope of this PR, let me try the previous direction and focus on fixing the SQL options propagation issue.

> I think assert as @cloud-fan suggest makes sense

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala [hunk 4]
```diff
@@ -87,31 +92,44 @@ object V2Writes extends Rule[LogicalPlan] with PredicateHelper {
...
       val newQuery = DistributionAndOrderingUtils.prepareQuery(deltaWrite, query, r.funCatalog)
       wd.copy(write = Some(deltaWrite), query = newQuery)
   }
 
+  private def mergeOptions(
+      commandOptions: Map[String, String],
+      dsOptions: Map[String, String]): Map[String, String] = {
+    // for DataFrame API cases, same options are carried by both Command and DataSourceV2Relation
+    // for DataFrameV2 API cases, options are only carried by Command
+    // for SQL cases, options are only carried by DataSourceV2Relation
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:128]: we will take care of `SessionConfigSupport` in followup PRs?

Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala:128]: it already takes care of `SessionConfigSupport` for DataFrame API (both v1 and v2), but not SQL. I will investigate the SQL cases later, but that may need some refactorings.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+      val insertResult = sql(s"SELECT * FROM $t1")
+      checkAnswer(insertResult, Seq(Row(1, "a"), Row(2, "b")))
+    }
+  }
+
+  test("SPARK-50286: Propagate options for DataFrame Append") {
+    val t1 = s"${catalogAndNamespace}table"
+    withTable(t1) {
+      sql(s"CREATE TABLE $t1 (id bigint, data string)")
+      withListener {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala:127]: we can use this new util function: https://github.com/apache/spark/pull/48867/files#diff-33334296c47a10ab23e8f3e62fc095aa04eb24227e06da99abb39769974bd41dR452

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        sql("VALUES (1, 'a'), (2, 'b')")
+          .withColumnRenamed("col1", "id")
+          .withColumnRenamed("col2", "data")
+          .writeTo(t1)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala:131]: `Seq(1 -> "a", 2 -> "b").toDF("id", "data")` is simpler

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    val t1 = s"${catalogAndNamespace}table"
+    withTable(t1) {
+      sql(s"CREATE TABLE $t1 (id bigint, data string)")
+      sql(s"INSERT INTO $t1 VALUES (1, 'a'), (2, 'b')")
+
+      withListener {
+        sql("VALUES (3, 'c'), (4, 'd')")
+          .withColumnRenamed("col1", "id")
+          .withColumnRenamed("col2", "data")
+          .writeTo(t1)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala:191]: shall we also test the v1 `DataFrameWriter` API?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala (hunk 6 above)
Comment (pan3793) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala:131]: thanks, simplified

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala (hunk 7 above)
Comment (pan3793) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala:217]: sure, add two cases
- `SPARK-50286: Propagate options for DataFrameWriter Append`
- `SPARK-50286: Propagate options for DataFrameWriter Overwrite`

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala (hunk 5 above)
Comment (pan3793) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala]: nice, I modified it to capture QueryExecution and reused

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**